designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+
+In the most basic form, the debug configuration references a SAM Template file location, and a Resource within that file. Other execution parameters can be configured, but are optional.
+
+Debugging local lambda invokes and local api gateway invokes each require slightly different inputs. The `aws-sam` Debug Configuration uses different request types to accommodate these variations.
+
+These debug configurations are authored in a json file. The following Toolkit assistance is provided:
+
+-   autocompletion with descriptions is provided for `aws-sam` related fields
+    -   There is no autocompletion available for specific values in a configuration. For example, if a user types in the location of a SAM Template file, there is no filesystem-based autocompletion. The Debug Configuration validates the configuration and notifies of errant values when it is run.
+-   snippets to produce typical (or starter) `aws-sam` debug configurations
+-   when no launch.json file is present in a workspace, VS Code exposes functionality that allows users to request auto-generated Debug Configurations. In this situation, the toolkit generates an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` Resources detected within all SAM Templates located in the workspace.
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations)
+
+Standalone Lambda function handlers are not supported through Debug Configurations.
+
+### <a id="codelenses"></a> CodeLenses
+
+#### CodeLenses in SAM Template files
+
+The following CodeLenses appear above every template resource of type `AWS::Serverless::Function`:
+
+-   Run Locally - See below for details
+-   Debug Locally - See below for details
+-   Configure - allows the user to configure a limited set of arguments that are used with the Run and Debug CodeLenses
+    -   Anything that can be defined by the SAM Template would not be configurable in here
+    -   This covers aspects like input event, and SAM CLI related arguments
+-   Add Debug Configuration - This is a utility feature to produce a pre-filled Debug Configuration in `launch.json` based on the associated template resource
+
+When clicked, the Run and Debug CodeLenses locally invoke their associated Template Resource. The following takes place:
+
+-   The SAM Application is built
+-   The associated SAM Template resource is invoked, using configurations set with the Configure CodeLens
+-   (If the Debug CodeLens was clicked) The VS Code debugger is attached to the invoked resource
+
+When clicked, the Configure CodeLens opens a (JSON) configuration file that resides in the workspace and is read by the Toolkit. The configuration file is used for each SAM Template Resource within the workspace. Users have autocompletion support in this file. A rich UI for configuration is not considered at this time, but the door remains open to adding a visual editor in the future based on user feedback.
```

Comment:
These read odd, it seems you are now talking about each bullet point from above, but you dont find that out till the 4th word, and the paragraph above starts the same

Reply:
You're right, I will revamp this